### DISCUSSION TITLE
[DismissableLayer] why Escape key is handled on document level?

### DISCUSSION THREAD

**Discussion by misha-erm** [radix-ui/primitives#2644] [Help]
Hello, 

Recently I've found out that Escape key is handled seprately for components like DropdownMenu and event handler for it is attached to document instead of a Content element.

Imagine a scenario where DropdownMenu is rendered inside a Parent component which also listens to escape and closes itself. How would I properly stop event propagation and close only DropdownMenu since doing something like `<DropdownMenu.Content onKeyDown={(e)=>e.stopPropagation()}` breaks it completely ?

Thanks in advance for any tips

**Comment by jjenzz** [comment]
it's on the document level for a11y reasons, if a dropdown menu is inside a dialog, pressing escape should only close the dropdown and pressing it again wld close the dialog.

radix manages this *layering* for you but if it isn working, you've likely stumbled upon [this](https://github.com/radix-ui/primitives/issues/1088) bug https://github.com/radix-ui/primitives/issues/1088

**Reply by misha-erm** [reply]
unfortunately dropdown menu is not inside radix's dialog component, it's inside our own primitive Panel which is a regular div, neither portalled nor floating
<img width="1114" alt="image" src="https://github.com/radix-ui/primitives/assets/8783498/d217ba54-bac9-4a1e-8163-dc1f38c5cf4e">

The panel has `onKeyDown` handler which listens for "Esc". So when you press Esc with dropdown opened it first bubbles up to panel and never reaches the document.

Probably the way we've done it is not the best and that's us who should move the logic from div to document as well, but still I'd like to get a clear understanding of it. Shouldn't radix listen for Esc in capturing phase?

**Reply by jjenzz** [reply]
oh interesting! yes, on the surface it sounds like radix shld capture 🤔 feel free to create an issue for that. the team will know more—i've not been close to the project recently.

**Comment by benoitgrelard** [comment]
Closing, addressed in #2653
